internal/provider/data_source_fmc_hosts.go

```diff
@@ -0,0 +1,162 @@
+	resp.TypeName = req.ProviderTypeName + "_hosts"
+}
+
+func (d *HostsDataSource) Schema(ctx context.Context, req datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		// This description is used by the documentation generator and the language server.
+		MarkdownDescription: "This data source can read the Hosts.",
+
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
+				MarkdownDescription: "The id of the object",
+				Computed:            true,
+			},
+			"domain": schema.StringAttribute{
+				MarkdownDescription: "The name of the FMC domain",
+				Optional:            true,
+			},
+			"items": schema.MapNestedAttribute{
+				MarkdownDescription: "Map of hosts. The key of the map is the name of the individual Host. Renaming Hosts in bulk is not yet implemented.",
+				Optional:            true,
+				Computed:            true,
+				NestedObject: schema.NestedAttributeObject{
+					Attributes: map[string]schema.Attribute{
+						"id": schema.StringAttribute{
+							MarkdownDescription: "UUID of the managed Host.",
+							Computed:            true,
+						},
+						"description": schema.StringAttribute{
+							MarkdownDescription: "Optional user-created description.",
+							Computed:            true,
+						},
+						"overridable": schema.BoolAttribute{
+							MarkdownDescription: "Indicates whether object values can be overridden.",
+							Computed:            true,
+						},
+						"ip": schema.StringAttribute{
+							MarkdownDescription: "IP of the host.",
+							Computed:            true,
+						},
+						"type": schema.StringAttribute{
+							MarkdownDescription: "Type of the object; this value is always 'Host'.",
+							Computed:            true,
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func (d *HostsDataSource) Configure(_ context.Context, req datasource.ConfigureRequest, _ *datasource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	d.client = req.ProviderData.(*FmcProviderData).Client
+}
+
+// End of section. //template:end model
+
+func (d *HostsDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
+	var config Hosts
+
+	// Read config
+	diags := req.Config.Get(ctx, &config)
+	resp.Diagnostics.Append(diags...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Set request domain if provided
+	reqMods := [](func(*fmc.Req)){}
+	if !config.Domain.IsNull() && config.Domain.ValueString() != "" {
+		reqMods = append(reqMods, fmc.DomainName(config.Domain.ValueString()))
+	}
+
+	tflog.Debug(ctx, fmt.Sprintf("%s: Beginning Read", config.Id.String()))
+
+	// In one request, up to 1000 elements can be read
+	// Here all elements from FMC are requested and locally only required are processed
+	offset := 0
+	limit := 1000
+	fullOutput := `{"items":[]}`
+	for page := 1; ; page++ {
```

Comment:
As discussed, let's try to handle pagination transparently in the client.

Reply:
Implementation moved to go-fmc.